### PR TITLE
authorizePublish gets called before authenticate finishes

### PR DESCRIPTION
**Pull request by jdiamond**
Hi @mcollina: what do you think of this test? I can try to fix it if it's legit and you have any hints on where I should look. Thanks.

### CODE REVIEW

**Comment by jdiamond** [comment]
My commit was a failing test to demonstrate the bug. Guess I should have expected Travis would fail when I had to bypass the pre-commit hooks...

**Comment by mcollina** [comment]
@jdiamond I think we should somehow stop parsing of new packets while processing a CONNECT (see https://github.com/mcollina/aedes/blob/master/lib/client.js).

**Comment by jdiamond** [comment]
Don't know if my commit is a good way to do the pausing, but it does make my test pass. I also pushed my branch to our QA environment where I first saw this bug and I'm not seeing it anymore.

Travis says there is a failing test in node 7. Not sure why, though. I just ran all tests 5 times in a row using node 7 and never saw any failures.

File: lib/client.js [hunk 1]
```diff
@@ -62,6 +66,17 @@ function Client (broker, conn) {
 
   nextBatch()
 
+  function pause () {
+    that._paused = true
+  }
+  this._pause = pause
```
Comment (mcollina) [inline lib/client.js]: can you define this a prototype method?

File: lib/client.js [hunk 2]
```diff
@@ -62,6 +66,17 @@ function Client (broker, conn) {
...
+
+  function unpause () {
+    that._paused = false
+    nextBatch()
+  }
```
Comment (mcollina) [inline lib/client.js]: this as well.

File: lib/handlers/connect.js [hunk 3]
```diff
@@ -14,14 +14,26 @@ function ClientPacketStatus (client, packet) {
...
 
+function pauseParsing (arg, done) {
+  this.client._pause()
+  done()
+}
+
+function resumeParsing (arg, done) {
+  this.client._unpause()
+  done()
+}
```
Comment (mcollina) [inline lib/handlers/connect.js]: These are sync steps, I will try to embed them in others, i.e. authenticate and emptyQueue.

**Comment by jdiamond** [comment]
Updated per your comments.

Same test is failing, but in node 4 this time. I still can't see how it's related to my change, though.

**Comment by mcollina** [comment]
Thanks!!

**Comment by jdiamond** [comment]
Thanks for the quick publish. Behaving great now. Next stop, production. 😄

**Pull request merged**